ir/visitor.cpp

```diff
@@ -20,76 +20,68 @@ limitations under the License.
 
 class Visitor::ChangeTracker {
     struct visit_info_t {
-        bool visit_finished;
+        bool visit_in_progress;
         const IR::Node *result;
     };
-    // typedef unordered_map<const IR::Node *, std::pair<bool, const IR::Node *>>  visited_t;
     typedef unordered_map<const IR::Node *, visit_info_t>  visited_t;
     visited_t           visited;
 
  public:
-    struct change_t {
-        // std::pair<visited_t::iterator, bool>    state;  // result of visited.emplace
-        bool           valid;
-        bool           already_present;
-        const IR::Node *original_node;
-        visit_info_t   *visit_info;    // Points into ChangeTracker.visited.
-
-        change_t() : valid(false) {}
-        change_t(visited_t *visited, const IR::Node *n) : valid(true) {
-                // Initialization
-                visited_t::iterator visited_it;
-                bool inserted;
-                std::tie(visited_it, inserted) =
-                    visited->emplace(n, visit_info_t({false,n}));
-                already_present = !inserted;
-                original_node = visited_it->first;
-                visit_info = &(visited_it->second);
-
-                // Sanity check for IR loops
-                if (!inserted && !visit_info->visit_finished)
-                    BUG("IR loop detected "); }
-
-        explicit operator bool() { return valid; }
-        bool done() { return already_present; }
-        const IR::Node *orig() { return original_node; }
-        const IR::Node *result() { return visit_info->result; }
-    };
 
-    bool done(const IR::Node *n) const {
-        return visited.count(n) && visited.at(n).visit_finished;
+    /* Begin tracking `n` during a visiting pass.  Use `finish` to mark `n` as
```

Comment:
For this to be a Doxygen comment, you just need to use `/**` at the beginning of the comment, like so:

```
/**
 * Begin tracking (...)
 */
```